engine/runtime/src/main/java/org/enso/interpreter/runtime/ResourceManager.java

```diff
@@ -0,0 +1,268 @@
+   * @param resource the resource to finalize.
+   */
+  @CompilerDirectives.TruffleBoundary
+  public void close(ManagedResource resource) {
+    Item it = items.remove(resource.getPhantomReference());
+    if (it == null) {
+      return;
+    }
+    // Unconditional finalization – user controls the resource manually.
+    it.doFinalize(context);
+  }
+
+  /**
+   * Removes the resource from the system, cancelling any automatic finalization action attached to
+   * it.
+   *
+   * @param resource the resource to take away from this system.
+   */
+  @CompilerDirectives.TruffleBoundary
+  public void take(ManagedResource resource) {
+    items.remove(resource.getPhantomReference());
+  }
+
+  private void tryFinalize(Item it) {
+    if (it.isFlaggedForFinalization().get()) {
+      if (it.getParkedCount().get() == 0) {
+        // We already know that isFlaggedForFinalization was true at some
+        // point and there are no other threads still parking the underlying
+        // value. Note that it is impossible for parked count to increase after
+        // the value is flagged for finalization, as parking the value requires
+        // a live reference. We need to check if another thread didn't reach
+        // here earlier to perform the finalization and reset the flag, so that
+        // no further attempts are made.
+        boolean continueFinalizing = it.isFlaggedForFinalization().compareAndSet(true, false);
+        if (continueFinalizing) {
+          it.doFinalize(context);
+          items.remove(it.reference);
+        }
+      }
+    }
+  }
+
+  /**
+   * Registers a new resource to the system. {@code function} will be called on {@code object} when
+   * the value returned by this method becomes unreachable.
+   *
+   * @param object the underlying resource
+   * @param function the finalizer action to call on the underlying resource
+   * @return a wrapper object, containing the resource and serving as a reachability probe
+   */
+  public ManagedResource register(Object object, Object function) {
+    if (isClosed) {
+      throw new IllegalStateException("Can't register new resources after context is closed.");
+    }
+    if (workerThread == null || !workerThread.isAlive()) {
+      worker.setKilled(false);
+      workerThread = context.createThread(worker);
+      workerThread.start();
+    }
+    ManagedResource resource = new ManagedResource(object);
+    PhantomReference<ManagedResource> ref = new PhantomReference<>(resource, referenceQueue);
+    resource.setPhantomReference(ref);
+    items.put(ref, new Item(object, function, ref));
+    return resource;
+  }
+
+  /**
+   * Stops this system, stops and joins any threads created by it. Unconditionally finalizes all the
+   * registered resources, ignoring their reachability status.
+   *
+   * <p>This is only useful when the underlying context is being finalized and no more user code
+   * will be run in it.
+   */
+  public void shutdown() {
+    isClosed = true;
```

Comment:
`isClosed` should probably be an atomic bool for safety here. Otherwise you can potentially race on `register` and `shutdown`.

Reply:
There's no race, `shutdown` must be called after all other threads are killed.